### ISSUE TITLE
If use different model structure to fine tune, wired

### ISSUE THREAD

**Issue by testpppppp** [songmzhang/DSKD#23]
Hi Zhang,
 I use your work recently, but find the loss drop while when decoding the result is some span repeat. not meet my expectation. 
I use the 0.5 * original loss + 0.5*(t2s_celoss +t2s_kdloss+s2t_kdloss). and for some reason there is lots token cannot concat
and I find the error occur when I use t2s_kdloss since it change student.logits directly.
so have you found the t2s_weight make too much token mapping to same token?
The following is some of my  codes

![image](https://github.com/user-attachments/assets/b7a7b533-f4fb-4898-8a11-55280c11c84f)

```
stu_q_hiddens = self.cma_query(stu_target_embeds).float()
tea_k_hiddens = norm_tea_target_embeds.float()
stu_v_hiddens = self.student_mlp(student_hiddens).float()
tea_v_hiddens = self.teacher_mlp(norm_teacher_hiddens + norm_tea_target_embeds).float()
t2s_kd_loss = self.dist_func(
                outputs.logits, t2s_logits.detach(), target, reduction="none", use_tea_temp=True
            )
```
 loss
![image](https://github.com/user-attachments/assets/5c751907-63e4-4485-83c8-30b48340acab)

**Comment by songmzhang** [comment]
It seems that this error didn't happen to me. Could you provide more detailed information about this error? For example:
1. Which models did you use for the student and the teacher model?
2. What is the student_outputs.loss/teacher_outputs.loss (did you pass the labels to model.forward()?)? If it stands for the NLL loss on the ground-truth labels, this value is somewhat large.
3. Why teacher_outputs.loss also drops? Did you also train the teacher model while distillation?
4. Are the labels correct? i.e., left shift 1 position based on input_ids

Theoretically, when t2s_ce_loss has converged to 0.0x, it means that the t2s_probs have high accuracy on predicting the student's target tokens, so t2s_kdloss should not have much negative effect.

**Comment by testpppppp** [comment]
Thank for you reply.
1. teacher is Mistral and student is gemma2
2. It's a caption model,  backbone is llava, and input include some VIDEO_TOKEN, so the input and output not shift, so the q and k is just caption target, the size not same and I pad -100 to the hidden length.
3. It's the fluctuation of teacher.loss , I have set no_grad and eval() for teacher_model
4. If I just keep student.loss and t2s_ce_loss, the output is normal(simply comment out the other)
5. the t2s weight somehow same.
![image](https://github.com/user-attachments/assets/c565d3ac-74d9-487f-801d-ebccc772e994)

6. Agree with you "Theoretically, when t2s_ce_loss has converged to 0.0x, it means that the t2s_probs have high accuracy on predicting the student's target tokens, so t2s_kdloss should not have much negative effect."

Let me think again.

**Comment by songmzhang** [comment]
Thanks for more information. According to the figure you post, it seems that the t2s_weight is overly smoothed and indeed abnormal. Please make sure that the visual and prompt tokens (input tokens) are not involved in the attention of the caption tokens (output tokens), since the attention is only needed for output tokens. And you can also check the predictions (argmax) of t2s_logits to see whether there is anything incorrect.
